Support recursive forward references in `from_type()` (#4614)
HypothesisWorks/hypothesis: hypothesis/RELEASE.rst

> RELEASE_TYPE: patch
> 
> This patch adds support for recursive forward references in
> :func:`~hypothesis.strategies.from_type`, such as
> ``A = list[Union["A", str]]`` (:issue:`4542`).
> Previously, such recursive type aliases would raise a ``ResolutionFailed``
> error. Now, Hypothesis can automatically resolve the forward reference
> by looking it up in the caller's namespace. This also resolves forward
> references inside ``type[...]``, such as ``type["MyClass"]``.

Reviewer: ````suggestion
RELEASE_TYPE: patch

This patch adds support for resolving forward references in |st.from_type|  (:issue:`4542`):

```python
# this now works
A = list[Union["A", str]]
st.from_type(A).example() 

# this also now works
s = st.from_type(list["B"])
@dataclass
class B:
    v: int
s.example()
```

Previously, these would raise an error. Now, Hypothesis automatically resolves the forward reference by looking it up in the caller's namespace.
